adapters/play/src/main/scala/caliban/PlayWSMessage.scala

```diff
@@ -0,0 +1,27 @@
+package caliban
+
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+import play.api.mvc.WebSocket.MessageFlowTransformer
+
+case class PlayWSMessage(messageType: String, id: Option[String] = None, payload: Option[JsValue] = None) {
+  lazy val request: Option[GraphQLRequest] = payload.flatMap(_.asOpt[GraphQLRequest])
+}
+
+object PlayWSMessage {
+  def apply[T](messageType: String, id: Option[String], payload: T)(implicit wr: Writes[T]): PlayWSMessage =
+    PlayWSMessage(messageType, id, Some(wr.writes(payload)))
+
+  implicit val playWSMessageReads: Reads[PlayWSMessage] =
+    ((__ \ "type").read[String] and
+      (__ \ "id").readNullable[String]
+      and (__ \ "payload").readNullable[JsValue])(PlayWSMessage.apply _)
+
+  implicit val playWSMessageWrites: Writes[PlayWSMessage] =
+    ((JsPath \ "type").write[String] ~
+      (JsPath \ "id").writeNullable[String] ~
+      (JsPath \ "payload").writeNullable[JsValue])(unlift(PlayWSMessage.unapply))
```

Comment:
Can collapse these `Reads` and `Writes` into a `Format`

either: 
`implicit val playWSMessageFormat: Format[PlayWSMessage] = Json.format[PlayWSMessage]`

or if the `macro` won't work because there are two apply methods then can do:

```scala
  implicit val playWSMessageReads: Reads[PlayWSMessage] =
    ((__ \ "type").format[String] and
      (__ \ "id").formatNullable[String]
      and (__ \ "payload").formatNullable[JsValue])(PlayWSMessage.apply _, unlift(PlayWSMessage.unapply))
```

Reply:
done!